src/main/scala/mesosphere/marathon/api/serialization/ContainerSerializer.scala

```diff
@@ -2,49 +2,49 @@ package mesosphere.marathon.api.serialization
 
 import mesosphere.marathon.Protos
 import mesosphere.marathon.core.externalvolume.ExternalVolumes
-import mesosphere.marathon.state.Container.Docker.PortMapping
+import mesosphere.marathon.state.Container.DockerDocker.PortMapping
 import mesosphere.marathon.state._
 import org.apache.mesos
 
 import scala.collection.JavaConverters._
 import scala.collection.immutable.Seq
 
 object ContainerSerializer {
+  def fromProto(proto: Protos.ExtendedContainerInfo): Container = {
+    if (proto.hasDocker) {
+      DockerDockerSerializer.fromProto(proto)
+    } else if (proto.hasMesosDocker) {
+      MesosDockerSerializer.fromProto(proto)
+    } else if (proto.hasMesosAppC) {
+      MesosAppCSerializer.fromProto(proto)
+    } else {
+      Container.Mesos(
+        volumes = proto.getVolumesList.asScala.map(Volume(_)).to[Seq]
+      )
+    }
+  }
+
   def toProto(container: Container): Protos.ExtendedContainerInfo = {
     val builder = Protos.ExtendedContainerInfo.newBuilder
       .addAllVolumes(container.volumes.map(VolumeSerializer.toProto).asJava)
 
     container match {
       case _: Container.Mesos =>
         builder.setType(mesos.Protos.ContainerInfo.Type.MESOS)
-      case docker: Container.Docker =>
+      case dd: Container.DockerDocker =>
```

Comment:
the `trait Container` is not sealed, therefore the compiler cannot check whether this check is exhaustive. Please change Container to be a `sealed trait`.